apps/devhub/utils.py

```diff
@@ -84,3 +97,203 @@ def escape_validation(validation):
+        # version which has been signed.
+        files = (File.objects.filter(Q(version__addon=self.addon) &
+                                     (Q(status__in=statuses) |
+                                      Q(status=amo.STATUS_BETA,
+                                        is_signed=True)))
+                 .order_by('-id'))
+
+        if self.file:
+            # Add some extra filters if we're validating a File instance,
+            # to try to get the closes possible match.
+            files = (files.exclude(pk=self.file.pk)
+                     # Files which are not for the same platform, but have
+                     # other files in the same version which are.
+                     .exclude(~Q(platform=self.file.platform) &
+                              Q(version__files__platform=self.file.platform))
+                     # Files which are not for the either the same platform or
+                     # for all platforms, but have other versions in the same
+                     # version which are.
+                     .exclude(~Q(platform__in=(self.file.platform,
+                                               amo.PLATFORM_ALL.id)) &
+                              Q(version__files__platform=amo.PLATFORM_ALL.id)))
+
+        for file in files:
+            # Only accept versions which come before the one we're validating.
+            if Version(file.version.version) < version:
+                return file
+
+
+class ValidationComparator(object):
+    """Compares the validation results from an older version with a version
+    currently being checked, and annotates results based on which messages
+    may be ignored for the purposes of automated signing."""
+
+    def __init__(self, validation):
+        self.validation = validation
+
+        self.messages = {self.message_key(msg): msg
+                         for msg in validation['messages']
+                         if 'context' in msg}
+
+    @classmethod
+    def message_key(cls, message):
+        """Returns a hashable key for a message based on properties
+        required when searching for a match."""
+
+        # No context, message is not matchable.
+        if 'context' not in message:
+            return None
+
+        return (tuple(message['id']),
+                tuple(message['context']),
+                message['file'],
+                message.get('signing_severity'),
+                ('context_data' in message and
+                 tuple(message['context_data'].iteritems())))
+
+    @classmethod
+    def match_messages(cls, message1, message2):
+        """Returns true if the two given messages match. The two messages
+        are assumed to have identical `message_key` values."""
+
+        # We'll eventually have to deal with issues like line number
+        # matching, with some kind of slop factor. For now, just rely
+        # on the basic heuristic of file name and context data.
+        return True
```

Comment:
Are the line numbers included in the message key at the moment? It's not clear from the comment if at the moment it's "too strict" (only exact line numbers will match) or "too lax" (we don't care about the line numbers at all).